From 3dc3ab3cee9c21c36a34a9dc9d3f099b6be4f153 Mon Sep 17 00:00:00 2001
From: zhouzilong <zhouzilong@uniontech.com>
Date: Thu, 12 Mar 2026 16:49:01 +0800
Subject: [PATCH] fix swig_python_appendoutput compilation argument error

---
 src/selinuxswig_python.i | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/selinuxswig_python.i b/src/selinuxswig_python.i
index 17e03b9..03ed296 100644
--- a/src/selinuxswig_python.i
+++ b/src/selinuxswig_python.i
@@ -71,7 +71,7 @@ def install(src, dest):
 	for (i = 0; i < *$2; i++) {
 		PyList_SetItem(list, i, PyString_FromString((*$1)[i]));
 	}
-	$result = SWIG_Python_AppendOutput($result, list);
+	$result = SWIG_AppendOutput($result, list);
 }
 
 /* return a sid along with the result */
@@ -108,7 +108,7 @@ def install(src, dest):
 		plist = PyList_New(0);
 	}
 
-	$result = SWIG_Python_AppendOutput($result, plist);
+	$result = SWIG_AppendOutput($result, plist);
 }
 
 /* Makes functions in get_context_list.h return a Python list of contexts */
-- 
2.20.1

